core/src/banking_stage/consumer.rs

```diff
@@ -470,33 +470,40 @@ impl Consumer {
         chunk_offset: usize,
         pre_results: impl Iterator<Item = Result<(), TransactionError>>,
     ) -> ProcessTransactionBatchOutput {
+        // Lock accounts so that other threads cannot encode transactions that
+        // will modify the same account state
+        let (mut batch, lock_us) =
+            measure_us!(bank.prepare_sanitized_batch_with_results(txs, pre_results));
+
+        // After locking, select the transactions that can fit into the block
         let (
             (transaction_qos_cost_results, cost_model_throttled_transactions_count),
             cost_model_us,
         ) = measure_us!(self.qos_service.select_and_accumulate_transaction_costs(
             bank,
-            txs,
-            pre_results
+            batch.sanitized_transactions(),
+            batch.lock_results().iter().cloned(),
         ));
 
-        // Only lock accounts for those transactions are selected for the block;
-        // Once accounts are locked, other threads cannot encode transactions that will modify the
-        // same account state
-        let (batch, lock_us) = measure_us!(bank.prepare_sanitized_batch_with_results(
-            txs,
-            transaction_qos_cost_results.iter().map(|r| match r {
-                Ok(_cost) => Ok(()),
-                Err(err) => Err(err.clone()),
-            })
-        ));
+        // Unlock accounts for any transactions that were not selected for the block;
+        batch.unlock_failures(
```

Comment:
I doubt it outweighs the benefits of this, but did you measure the impact of this change on performance when we have many small batches? Maybe run bench-tps as a sanity check on this, using --num-conflict-groups 1

We're grabbing the account locks lock 3 times instead of twice per batch now, so could lead to more contention for that lock.